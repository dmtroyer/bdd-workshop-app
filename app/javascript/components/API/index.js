import axios from 'axios';

const fetchUserProfile = async id => {
  const response = await axios
    .get(`/api/v1/profiles/${id}`, {
      // TODO: set derault for all calls
      headers: { Accept: 'application/json' },
    })
    .catch(() => {
      return createUserProfile();
    });
  return response;
};

const createUserProfile = async () => {
  const response = await axios.post(
    '/api/v1/profiles',
    {},
    {
      headers: { Accept: 'application/json' },
    }
  );
  return response;
};

const updateUserProfile = async ({ data: { handle, id } }) => {
  const response = await axios.put(
    `/api/v1/profiles/${id}`,
    { player: { handle } },
    {
      headers: { Accept: 'application/json' },
    }
  );
  return response;
};

export { fetchUserProfile };
export { createUserProfile };
export { updateUserProfile };
export default { fetchUserProfile, createUserProfile, updateUserProfile };
