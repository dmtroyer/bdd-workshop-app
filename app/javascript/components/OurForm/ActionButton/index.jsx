import React from 'react';
import { Button } from 'reactstrap';
import PropTypes from 'prop-types';

const ActionButton = ({ children, isSubmitting }) => (
  <Button name="submit" color="primary" type="submit" disabled={isSubmitting} className="float-right">
    {children}
  </Button>
);

ActionButton.propTypes = {
  children: PropTypes.node,
  isSubmitting: PropTypes.bool,
};

export default ActionButton;
