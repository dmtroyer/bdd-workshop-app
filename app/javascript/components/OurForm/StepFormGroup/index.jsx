import React from 'react';
import { FormGroup, Input, Label } from 'reactstrap';
import PropTypes from 'prop-types';

const StepFormGroup = ({
  step,
  label,
  type,
  placeholder,
  values,
  handleChange,
}) => (
  <FormGroup>
    <Label for={step}>{label}</Label>
    <Input
      type={type || 'text'}
      name={step}
      value={values[step]}
      onChange={handleChange}
      placeholder={placeholder}
    />
  </FormGroup>
);
StepFormGroup.propTypes = {
  step: PropTypes.string,
  label: PropTypes.string,
  type: PropTypes.string,
  placeholder: PropTypes.string,
  values: PropTypes.object,
  handleChange: PropTypes.func,
};

export default StepFormGroup;
