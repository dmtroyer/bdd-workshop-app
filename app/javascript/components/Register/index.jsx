import React from 'react';
import MainContainer from '../MainContainer';
import OurForm from '../OurForm';
import messageBus from '../../utils/messageBus';
import { createUserProfile } from '../API';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

const Register = ({ fetchProfile, history }) => {
  const onOurFormSubmit = OurFormData => {
    return createUserProfile(OurFormData)
      .then(({ data: { id } }) => {
        localStorage.setItem('player', JSON.stringify({ id: id }));
        messageBus.info('profile successfully created');
        fetchProfile();
        history.push('/profile');
      })
      .catch(res => {
        const messages = Object.keys(res.response.data.errors)
          .map(key => [key, res.response.data.errors[key]].join(': '))
          .join();
        messageBus.error(messages);
      });
  };

  return (
    <MainContainer dataTestId="register">
      <h3>Please sign in or create a profile!</h3>
      <OurForm onSubmit={onOurFormSubmit} step="handle" />
      <Link className="btn" to="/sign_in">
        sign in with an existing account
      </Link>
    </MainContainer>
  );
};

Register.propTypes = {
  fetchProfile: PropTypes.func,
  history: PropTypes.object,
};

export default Register;
