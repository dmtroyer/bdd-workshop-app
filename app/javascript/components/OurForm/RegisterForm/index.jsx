import React from 'react';
import { Form } from 'reactstrap';
import PropTypes from 'prop-types';
import ActionButton from '../ActionButton';
import StepFormGroup from '../StepFormGroup';

const RegisterForm = props => (
  <Form onSubmit={props.handleSubmit}>
    {(props.step === 'handle' || !props.step) && (
      <StepFormGroup
        {...props}
        step="handle"
        label="Handle"
        placeholder="input a custom handle"
      />
    )}
    {(props.step === 'email' || !props.step) && (
      <StepFormGroup
        {...props}
        step="email"
        label="Email"
        type="email"
        placeholder="input your email"
      />
    )}
    <ActionButton>{props.step ? 'Next' : 'Submit'}</ActionButton>
  </Form>
);
RegisterForm.propTypes = {
  handleSubmit: PropTypes.func,
  step: PropTypes.string,
};

export default RegisterForm;
