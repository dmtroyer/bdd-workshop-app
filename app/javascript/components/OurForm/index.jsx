import React from 'react';
import { Formik } from 'formik';
import PropTypes from 'prop-types';
import RegisterForm from './RegisterForm';

const OurForm = ({ profile, onSubmit }) => {
  const { handle, email, avatarUrl } = profile;
  const step = !handle
    ? 'handle'
    : !email
    ? 'email'
    : !avatarUrl
    ? 'avatar'
    : undefined;
  return (
    <Formik
      initialValues={{ handle, email, avatarUrl }}
      onSubmit={(values, { setSubmitting }) => {
        setSubmitting(true);
        event.preventDefault();
        onSubmit(values).then(() => {
          setSubmitting(false);
        });
      }}
    >
      {props => <RegisterForm step={step} {...props} />}
    </Formik>
  );
};

OurForm.defaultProps = {
  profile: { handle: '', email: '', avatarUrl: '' },
};

OurForm.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  profile: PropTypes.shape({
    id: PropTypes.string,
    handle: PropTypes.string,
    email: PropTypes.string,
    avatarUrl: PropTypes.string,
  }),
};

export default OurForm;
