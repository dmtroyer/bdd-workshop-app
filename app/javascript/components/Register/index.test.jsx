import React from 'react';
import { shallow } from 'enzyme';
import Register from '.';
import OurForm from '../OurForm';
import { createUserProfile } from '../API';

jest.mock('../API');

describe('Register', () => {
  it('renders text and a OurForm', () => {
    const wrapper = shallow(<Register />);
    expect(wrapper).toMatchSnapshot();
  });

  it('will call the onSubmit method, WHEN the OurForm is submitted', () => {
    createUserProfile.mockResolvedValue({});
    const wrapper = shallow(<Register />);
    wrapper.find(OurForm).prop('onSubmit')({ data: { handle: 'the-handle' } });
    expect(createUserProfile).toHaveBeenCalledTimes(1);
    expect(createUserProfile).toHaveBeenCalledWith({
      data: { handle: 'the-handle' },
    });
  });
});
