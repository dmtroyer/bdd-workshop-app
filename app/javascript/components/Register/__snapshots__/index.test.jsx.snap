// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register renders text and a OurForm 1`] = `
<MainContainer
  dataTestId="register"
>
  <Row
    tag="div"
  >
    <Col
      md={
        Object {
          "offset": 3,
          "size": 6,
        }
      }
      sm="12"
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Card
        tag="div"
      >
        <CardHeader
          tag="div"
        >
          <h1>
            Register
          </h1>
        </CardHeader>
        <CardBody
          tag="div"
        >
          <CardTitle
            tag="div"
          >
            <span>
              Register or 
            </span>
            <Link
              replace={false}
              to="/sign_in"
            >
              sign in with an existing account
            </Link>
          </CardTitle>
          <OurForm
            onSubmit={[Function]}
            profile={
              Object {
                "avatarUrl": "",
                "email": "",
                "handle": "",
              }
            }
            step="handle"
          />
        </CardBody>
      </Card>
    </Col>
  </Row>
</MainContainer>
`;
